src/core/thread/dua_manager.cpp

```diff
@@ -75,23 +76,98 @@ void DuaManager::UpdateDomainUnicastAddress(BackboneRouter::Leader::DomainPrefix
 
     OT_ASSERT(prefix != NULL);
 
-    mDomainUnicastAddress.GetAddress().SetPrefix(prefix->mPrefix.mFields.m8, prefix->mLength);
     mDomainUnicastAddress.mPrefixLength = prefix->mLength;
+    mDomainUnicastAddress.GetAddress().Clear();
+    mDomainUnicastAddress.GetAddress().SetPrefix(prefix->mPrefix.mFields.m8, prefix->mLength);
 
-    if (Get<Utils::Slaac>().GenerateIid(mDomainUnicastAddress, NULL, 0, &mDadCounter) == OT_ERROR_NONE)
+    // Apply cached DUA Interface Identifier manually specified.
+    if (!mFixedDuaInterfaceIdentifier.IsUnspecified())
     {
-        Get<ThreadNetif>().AddUnicastAddress(mDomainUnicastAddress);
+        mDomainUnicastAddress.GetAddress().SetIid(mFixedDuaInterfaceIdentifier);
     }
     else
     {
-        mDomainUnicastAddress.GetAddress().Clear();
-        otLogWarnCore("Failed to generate valid DUA");
+        SuccessOrExit(GenerateDomainUnicastAddressIid());
     }
 
+    Get<ThreadNetif>().AddUnicastAddress(mDomainUnicastAddress);
+
 exit:
     return;
 }
 
+otError DuaManager::GenerateDomainUnicastAddressIid(void)
+{
+    otError error;
+
+    if ((error = Get<Utils::Slaac>().GenerateIid(mDomainUnicastAddress, NULL, 0, &mDadCounter)) == OT_ERROR_NONE)
+    {
+        otLogInfoCore("Generated DUA: %s", mDomainUnicastAddress.GetAddress().ToString().AsCString());
+    }
+    else
+    {
+        otLogWarnCore("Failed to generate valid DUA");
+    }
+
+    return error;
+}
+
+otError DuaManager::SetFixedDuaInterfaceIdentifier(const Ip6::InterfaceIdentifier *aIid)
```

Comment:
Suggestion here:
How about adding an new method `ClearFixedDuaInterfaceIdentifier()` to handle the clear code.
And then we can make `SetFixedDuaInterfaceIdentifier()` use a reference `const Ip6::InterfaceIdentifier &aIid` instead of a pointer,

Thoughts?

Reply:
@abtink do you suggest adding a new method only in `DuaManager` class? or also applying to otAPI?

If the former, it seems to me that keeping existing methods might be more consistent with the corresponding ot APIs?
 
If the latter, there was some opposite comment in the previous implementation which set the address directly
https://github.com/openthread/openthread/pull/4900#discussion_r417046630 by simonlingoogle@
> We can just use otThreadSetFixedDomainUnicastAddress(instance, NULL) for this purpose ? 

Additionally there are similar usage in existing ot API (combining set/clear in one API, though the parameter is not a pointer)
https://github.com/openthread/openthread/blob/b9d348c32879e626540f90e0996eab7d5861c653/include/openthread/link.h#L658

---
Both the two options - combining Set/Clear API/method or not - seem fine to me, we may better to arrive some consensus regarding the preference.
Thoughts?
@abtink @jwhui @simonlingoogle

Reply:
I don't have a strong opinion for this. My intention was to to reduce the number of API functions. 

In the meantime, I think we should better follow existing conventions of OpenThread code base. 
There are APIs that set some fixed configurations in OpenThread, e.g. `otThreadSetLocalLeaderPartitionId`. 
`otThreadSetLocalLeaderPartitionId` uses one API for setting and clearing fixed partition ID. 

Thoughts ?

Reply:
Let me first say, it's all fine for me. No strong preference either way.

I was just suggesting this for the internal methods (not the public ot APIs) - my though process for this was: 
- to avoid using pointer and use reference variable instead when setting the value (avoid `*aIId`)
- break the implementation into two smaller methods (`clear` and `set`) - just for ease of reading the code. 

But neither point is actually essential. 

I agree it's better to have fewer public ot API, so using `NULL` pointer indicate clear action in public API  sounds good to me.